src/tiledbsoma_ml/scvi.py

```diff
@@ -0,0 +1,146 @@
+                        `["dataset_id", "assay", "suspension_type", "donor_id"]`.
+
+        batch_labels: Sequence[str], optional
+        List of possible values of the batch label, for mapping to label tensors. By default,
+                        this will be derived from the unique labels in the given query results (given
+                        `batch_column_names`), making the label mapping depend on the query. The `batch_labels`
+                        attribute in the `SCVIDataModule` used for training may be saved and here restored in
+                        another instance for a different query. That ensures the label mapping will be correct
+                        for the trained model, even if the second query doesn't return examples of every
+                        training batch label.
+
+        dataloader_kwargs: dict, optional
+        Keyword arguments passed to `tiledbsoma_ml.experiment_dataloader()`, e.g. `num_workers`.
+        """
+        super().__init__()
+        self.query = query
+        self.dataset_args = args
+        self.dataset_kwargs = kwargs
+        self.dataloader_kwargs = (
+            dataloader_kwargs if dataloader_kwargs is not None else {}
+        )
+        self.batch_column_names = (
+            batch_column_names
+            if batch_column_names is not None
+            else ["dataset_id", "assay", "suspension_type", "donor_id"]
+        )
+        self.batch_colsep = "//"
+        self.batch_colname = "scvi_batch"
+        # prepare LabelEncoder for the scVI batch label:
+        #   1. read obs DataFrame for the whole query result set
+        #   2. add scvi_batch column
+        #   3. fit LabelEncoder to the scvi_batch column's unique values
+        if batch_labels is None:
+            obs_df = (
+                self.query.obs(column_names=self.batch_column_names)
+                .concat()
+                .to_pandas()
+            )
+            self._add_batch_col(obs_df, inplace=True)
+            batch_labels = obs_df[self.batch_colname].unique()
+        self.batch_labels = batch_labels
+        self.batch_encoder = LabelEncoder().fit(self.batch_labels)
+
+    def setup(self, stage: str | None = None) -> None:
+        # Instantiate the ExperimentDataset with the provided args and kwargs.
+        self.train_dataset = ExperimentDataset(
+            self.query,
+            *self.dataset_args,
+            obs_column_names=self.batch_column_names,  # type: ignore[arg-type]
+            **self.dataset_kwargs,  # type: ignore[misc]
+        )
+
+    def train_dataloader(self) -> DataLoader:
+        return experiment_dataloader(
+            self.train_dataset,
+            **self.dataloader_kwargs,
+        )
+
+    def _add_batch_col(
+        self, obs_df: pd.DataFrame, inplace: bool = False
+    ) -> pd.DataFrame:
+        # synthesize a new column for obs_df by concatenating the self.batch_column_names columns
+        if not inplace:
+            obs_df = obs_df.copy()
```

Comment:
Since pandas is moving towards Copy-Over-Write logic soon. We may want to be proactive and remove this `inplace` behaviour

Reply:
hm, the two uses of this function (in this file) use `inplace=True` to just add the column to the existing DF, and skip the `copy()`. It doesn't use the deprecated Pandas `inplace` kwarg, and we'll want to support `pandas<3` for some years, so I would leave this as-is.

Or: how would you propose rewriting it (without an unnecessary `copy()`)?